SteamKit2/SteamKit2/Steam/Handlers/SteamAuthTicket/SteamAuthTicket.cs

```diff
@@ -0,0 +1,210 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using SteamKit2.Internal;
+
+namespace SteamKit2
+{
+    /// <summary>
+    /// This handler generates auth session ticket and handles it's verification by steam.
+    /// </summary>
+    public sealed partial class SteamAuthTicket : ClientMsgHandler
+    {
+        private readonly Dictionary<EMsg, Action<IPacketMsg>> dispatchMap;
+        private readonly ConcurrentQueue<byte[]> gameConnectTokens = new ConcurrentQueue<byte[]>();
+        private readonly ConcurrentDictionary<uint, List<CMsgAuthTicket>> ticketsByGame = new ConcurrentDictionary<uint, List<CMsgAuthTicket>>();
+
+        internal SteamAuthTicket()
+        {
+            dispatchMap = new Dictionary<EMsg, Action<IPacketMsg>>
+            {
+                { EMsg.ClientAuthListAck, HandleTicketAcknowledged },
+                { EMsg.ClientTicketAuthComplete, HandleTicketAuthComplete },
+                { EMsg.ClientGameConnectTokens, HandleGameConnectTokens },
+                { EMsg.ClientLogOnResponse, HandleLogOnResponse }
+            };
+        }
+
+        /// <summary>
+        /// Generate session ticket, and verify it with steam servers.
+        /// </summary>
+        /// <param name="appid">The appid to request the ticket of.</param>
+        /// <returns><c>null</c> if user isn't fully logged in, doesn't own the game, or steam deemed ticket invalid; otherwise <see cref="TicketInfo" /> instance.</returns>
+        public async Task<TicketInfo?> GetAuthSessionTicket( uint appid )
```

Comment:
Task returning method seems out of place for me looking at the rest of the codebase, but couldn't figure out how to do it better.

Reply:
async is fine.